mm2src/coins/eth.rs

```diff
@@ -293,11 +299,86 @@ pub enum EthCoinType {
     Erc20 { platform: String, token_addr: Address },
 }
 
+/// An alternative to `crate::PrivKeyBuildPolicy`, typical only for ETH coin.
+pub enum EthPrivKeyBuildPolicy {
+    IguanaPrivKey(IguanaPrivKey),
+    GlobalHDAccount(GlobalHDAccountArc),
+    #[cfg(target_arch = "wasm32")]
+    Metamask(MetamaskArc),
+}
+
+impl EthPrivKeyBuildPolicy {
+    /// Detects the `EthPrivKeyBuildPolicy` with which the given `MmArc` is initialized.
+    pub fn detect_priv_key_policy(ctx: &MmArc) -> MmResult<EthPrivKeyBuildPolicy, CryptoCtxError> {
+        let crypto_ctx = CryptoCtx::from_ctx(ctx)?;
+
+        match crypto_ctx.key_pair_policy() {
+            KeyPairPolicy::Iguana => {
+                // Use an internal private key as the coin secret.
+                let priv_key = crypto_ctx.mm2_internal_privkey_secret();
+                Ok(EthPrivKeyBuildPolicy::IguanaPrivKey(priv_key))
+            },
+            KeyPairPolicy::GlobalHDAccount(global_hd) => Ok(EthPrivKeyBuildPolicy::GlobalHDAccount(global_hd.clone())),
+        }
+    }
+}
+
+impl TryFrom<PrivKeyBuildPolicy> for EthPrivKeyBuildPolicy {
+    type Error = PrivKeyPolicyNotAllowed;
+
+    /// Converts `PrivKeyBuildPolicy` to `EthPrivKeyBuildPolicy`
+    /// taking into account that  ETH doesn't support `Trezor` yet.
+    fn try_from(policy: PrivKeyBuildPolicy) -> Result<Self, Self::Error> {
+        match policy {
+            PrivKeyBuildPolicy::IguanaPrivKey(iguana) => Ok(EthPrivKeyBuildPolicy::IguanaPrivKey(iguana)),
+            PrivKeyBuildPolicy::GlobalHDAccount(global_hd) => Ok(EthPrivKeyBuildPolicy::GlobalHDAccount(global_hd)),
+            PrivKeyBuildPolicy::Trezor => Err(PrivKeyPolicyNotAllowed::HardwareWalletNotSupported),
+        }
+    }
+}
+
+/// An alternative to `crate::PrivKeyPolicy`, typical only for ETH coin.
+#[derive(Clone)]
+pub enum EthPrivKeyPolicy {
+    KeyPair(KeyPair),
+    #[cfg(target_arch = "wasm32")]
+    Metamask(MetamaskWeak),
+}
+
+impl fmt::Debug for EthPrivKeyPolicy {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            EthPrivKeyPolicy::KeyPair(_) => write!(f, "KeyPair"),
+            #[cfg(target_arch = "wasm32")]
+            EthPrivKeyPolicy::Metamask(_) => write!(f, "Metamask"),
+        }
+    }
+}
+
+impl From<KeyPair> for EthPrivKeyPolicy {
+    fn from(key_pair: KeyPair) -> Self { EthPrivKeyPolicy::KeyPair(key_pair) }
+}
+
+impl EthPrivKeyPolicy {
+    pub fn key_pair(&self) -> Option<&KeyPair> {
+        match self {
+            EthPrivKeyPolicy::KeyPair(key_pair) => Some(key_pair),
+            #[cfg(target_arch = "wasm32")]
+            EthPrivKeyPolicy::Metamask(_) => None,
+        }
+    }
+
+    pub fn key_pair_or_err(&self) -> MmResult<&KeyPair, PrivKeyPolicyNotAllowed> {
+        self.key_pair()
+            .or_mm_err(|| PrivKeyPolicyNotAllowed::HardwareWalletNotSupported)
+    }
```

Comment:
I repeated the same methods from `lp_coins::PrivKeyPolicy`, so how about removing `EthPrivKeyPolicy::key_pair` method, but leaving `EthPrivKeyPolicy::key_pair_or_err` as it is? If renaming it to `key_pair`, it's also worth to rename `PrivKeyPolicy::key_pair_or_err`. But this will result in changes to files not related to this task.
What do you think?

Reply:
> how about removing `EthPrivKeyPolicy::key_pair` method, but leaving `EthPrivKeyPolicy::key_pair_or_err` as it is?

I think it's also fine.

Reply:
Done